lib/concurrency/Prefix_Refinement.thy

```diff
@@ -256,79 +389,267 @@ corollary matching_fragment_matching_tr_trivR:
+  apply (subst self_closed_def, clarsimp simp: rely_def rely_cond_Cons_eq)
+  apply (drule(2) self_closedD)
+  apply (fastforce simp: rely_cond_Cons_eq)
+  done
+
+lemma rely_env_closed:
+  "\<lbrakk>env_closed P s f;
+    \<And>xs s. \<lbrakk>P' xs s; rely_cond R s0 xs\<rbrakk> \<Longrightarrow> P xs s \<and> R (last_st_tr xs s0) s\<rbrakk>
+   \<Longrightarrow> env_closed P' s (rely f R s0)"
+  apply (subst env_closed_def, clarsimp simp: rely_def)
+  apply (drule_tac s'=s' in env_closedD, assumption)
+   apply simp
+  apply (clarsimp simp: image_def)
+  apply (fastforce intro: rely_cond_Cons rev_bexI)
+  done
+
+lemma rely_cond_mono:
+  "R \<le> R' \<Longrightarrow> rely_cond R \<le> rely_cond R'"
+  by (simp add: le_fun_def rely_cond_def split_def)
+
+lemma is_matching_fragment_add_rely:
+  "\<lbrakk>is_matching_fragment sr osr r ctr cres s0 AR s f; AR' \<le> AR\<rbrakk>
+   \<Longrightarrow> is_matching_fragment sr osr r ctr cres s0 AR' s (rely f AR' s0)"
+  apply (frule is_matching_fragment_Nil)
+  apply (clarsimp simp: is_matching_fragment_def prefix_closed_rely
+                        rely_self_closed)
+  apply (intro conjI)
+    apply (erule rely_env_closed)
+    apply (frule rely_cond_mono)
+    apply (simp add: le_fun_def rely_cond_Cons_eq)
+   apply (fastforce simp: rely_def)
+  apply (auto simp: rely_def)[1]
+  done
+
+named_theorems pfx_refn_rely_pre
+(* Introduce schematic rely relations, fail if already schematic *)
+method pfx_refn_rely_pre = WP_Pre.pre_tac pfx_refn_rely_pre
+
+lemma prefix_refinement_weaken_rely[pfx_refn_rely_pre]:
+  "\<lbrakk>prefix_refinement sr isr osr r AR R P Q f g; R' \<le> R; AR' \<le> AR\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr osr r AR' R' P Q f g"
+  apply (subst prefix_refinement_def, clarsimp)
+  apply (drule(3) prefix_refinementD, assumption+)
+  apply (clarsimp simp: rely_cond_def split_def le_fun_def)
+  apply (rule exI, rule conjI, erule is_matching_fragment_add_rely)
+   apply (simp add: le_fun_def)
+  apply (auto simp add: triv_refinement_def rely_def)
+  done
+
+
+section \<open>Inserting assumptions to be proved later\<close>
+
+lemma prefix_refinement_req:
+  "\<lbrakk>\<And>s0 s t0 t. \<lbrakk>sr s0 t0; isr s t; P s0 s; Q t0 t\<rbrakk> \<Longrightarrow> F;
+    F \<Longrightarrow> prefix_refinement sr isr osr rvr AR R P Q f g\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr osr rvr AR R P Q f g"
+  by (auto simp: prefix_refinement_def)
+
+\<comment> \<open>FIXME: is it more usable to have this lemma combined like this or to split it up like Corres_UL's version\<close>
+lemma prefix_refinement_gen_asm:
+  "\<lbrakk>P \<Longrightarrow> prefix_refinement sr isr osr rvr AR R P' Q' f g\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr osr rvr AR R (P' and (\<lambda>_ _. P)) Q' f g"
+  "\<lbrakk>P \<Longrightarrow> prefix_refinement sr isr osr rvr AR R P' Q' f g\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr osr rvr AR R P' (Q' and (\<lambda>_ _. P)) f g"
```

Comment:
Depends on how this is used. If it is used with instantiation, then having them separate is better. If it is used by matching against a current goal to extract `P`, then a single name is better. We could get both but providing two separate names and a `lemmas` statement for the collection.

Reply:
It looks like the `corres` version is used both ways, so I've split this up and added `lemmas prefix_refinement_gen_asms = ...` like you suggested.